### PR TITLE
fix: delegate starting receiver loops to caller

### LINKED ISSUE #68: When using as library, the dealer logs an error every 30 seconds about not having received a pong
I'm using go-librespot as a library in my own application.
I don't use the dealer, but it logs `did not receive last pong from dealer` every 30s.

It seems that the necessary receive loop is only started after requesting something from the dealer, but the send loop is always active.

I'll make a PR to fix this

### PR DESCRIPTION
**Pull request by silenium-dev**
fixes #68

### CODE REVIEW

**Comment by devgianlu** [comment]
The reason why the receiver isn't started is that there might be messages that arrive immediately after connection and if there's no listener then they'll be lost.

You'll need to call `ReceiveMessage` or `ReceiveRequest` and read from the channel they return. Or find an actual fix?

**Comment by silenium-dev** [comment]
Maybe we could start the ping loop on the first `ReceiveMessage`/`ReceiveRequest` call? Because it shouldn't be started either, if there's no receiver that could process the pings.
Or we even could open the websocket connection on the first call, instead of immediately on creation.

**Comment by devgianlu** [comment]
> Maybe we could start the ping loop on the first `ReceiveMessage`/`ReceiveRequest` call? Because it shouldn't be started either, if there's no receiver that could process the pongs.

That makes sense, but I believe that the server will close the connection if we don't ping it.

> Or we even could open the websocket connection on the first call, instead of immediately on creation.

This makes even more sense, but we must be careful to the fact that the connection will be nil.

Because connection can fail, I don't think delegating it to the `Receive*` methods is feasible. Instead it should probably be handled by the caller (outside of `NewDealer`) through an exported `Connect` method, but we must deal with nil `conn` pointers and the `reconnectLock` (which should probably be named `connMu`). 

All of the above also applies to the `Accesspoint`.

**Comment by silenium-dev** [comment]
I'm going to reopen this, as soon as I have implemented a proper fix

**Comment by silenium-dev** [comment]
I'm unsure whether to move `Accesspoint.Connect()` out of `Session` as `AudioKeyProvider` and `Spclient` depend on it.
Nothing internal depends on Dealer, so moving `Connect()` out of `Session` seems ok.

**Comment by silenium-dev** [comment]
It seems like the dealer ping timings are a bit too tight, it sends a ping every 30s, but also has a timeout of 30s, which can cause some network latency related reconnects

**Comment by silenium-dev** [comment]
Seems to be fixed by setting the time that the pong can be behind to `pingInterval + timeout`

File: dealer/dealer.go [hunk 1]
```diff
@@ -185,16 +212,16 @@ loop:
 		}
 	}
 
-	_ = d.conn.Close(websocket.StatusInternalError, "")
-
 	// if we shouldn't stop, try to reconnect
 	if !d.stop {
-		d.reconnectLock.Lock()
+		_ = d.conn.Close(websocket.StatusInternalError, "")
```
Comment (devgianlu) [inline dealer/dealer.go:217]: Why do this only when not stopping?

File: dealer/dealer.go [hunk 2]
```diff
@@ -149,8 +167,16 @@ loop:
 		case <-d.recvLoopStop:
 			break loop
 		default:
-			// no need to hold the reconnectLock since reconnection happens in this routine
+			// no need to hold the connMu since reconnection happens in this routine
```
Comment (devgianlu) [inline dealer/dealer.go:170]: This comment is in contrast with what happens below, is it required to lock for reading?

File: dealer/dealer.go [hunk 3]
```diff
@@ -83,20 +86,34 @@ func (d *Dealer) connect() error {
...
+	log.Debugf("dealer connection opened")
+
 	return nil
 }
 
 func (d *Dealer) Close() {
+	d.connMu.Lock()
+	defer d.connMu.Unlock()
+
+	if d.conn == nil {
```
Comment (devgianlu) [inline dealer/dealer.go:98]: I'd set `d.stop = true` anyway.

File: ap/ap.go [hunk 4]
```diff
@@ -189,15 +185,22 @@ func (ap *Accesspoint) Connect(creds *pb.LoginCredentials) error {
 }
 
 func (ap *Accesspoint) Close() {
+	ap.connMu.Lock()
+	defer ap.connMu.Unlock()
+
+	if ap.conn == nil {
```
Comment (devgianlu) [inline ap/ap.go:191]: I'd set `ap.stop = true` anyway.

**Review by devgianlu — CHANGES_REQUESTED** [review]

**Comment by devgianlu** [comment]
> I'm unsure whether to move `Accesspoint.Connect()` out of `Session` as `AudioKeyProvider` and `Spclient` depend on it.
> Nothing internal depends on Dealer, so moving `Connect()` out of `Session` seems ok.

I see, I guess this is ok because the accesspoint is required to complete authentication so it has to be always started.

> Seems to be fixed by setting the time that the pong can be behind to `pingInterval + timeout`. `lastPong` is now reset on reconnect to prevent a reconnect loop.

Sure, this makes sense.

File: dealer/dealer.go (hunk 1 above)
Comment (silenium-dev) [reply dealer/dealer.go:217]: Because `dealer.Close()` handles closing the connection when we are stopping

File: dealer/dealer.go (hunk 3 above)
Comment (silenium-dev) [reply dealer/dealer.go:100]: Changed

File: ap/ap.go (hunk 4 above)
Comment (silenium-dev) [reply ap/ap.go:193]: Changed

File: dealer/dealer.go (hunk 2 above)
Comment (silenium-dev) [reply dealer/dealer.go:171]: Removed the lock

**Comment by devgianlu** [comment]
Thanks!

**Comment by devgianlu** [comment]
Had to implement a couple of fixes, looks better now.

**Pull request merged**